airbyte-server/src/main/java/io/airbyte/server/handlers/helpers/CatalogConverter.java

```diff
@@ -30,7 +42,30 @@ private static io.airbyte.api.model.generated.AirbyteStream toApi(final io.airby
         .namespace(stream.getNamespace());
   }
 
-  private static io.airbyte.protocol.models.AirbyteStream toProtocol(final io.airbyte.api.model.generated.AirbyteStream stream) {
+  @SuppressWarnings("PMD.AvoidLiteralsInIfCondition")
+  private static io.airbyte.protocol.models.AirbyteStream toProtocol(final AirbyteStream stream, AirbyteStreamConfiguration config)
+      throws JsonValidationException {
+    JsonNode streamSchema = stream.getJsonSchema();
+    if (config.getFieldSelectionEnabled() != null && config.getFieldSelectionEnabled()) {
+      // Validate the selected field paths.
+      for (final var selectedFieldInfo : config.getSelectedFields()) {
+        if (selectedFieldInfo.getFieldPath().isEmpty()) {
+          throw new JsonValidationException("Selected field path cannot be empty");
+        }
+        if (selectedFieldInfo.getFieldPath().size() > 1) {
+          // TODO(mfsiega-airbyte): support nested fields.
+          throw new UnsupportedOperationException("Nested field selection not supported");
+        }
+      }
+      // Only include the selected fields.
+      List<String> selectedFieldNames = config.getSelectedFields().stream().map((field) -> field.getFieldPath().get(0)).collect(Collectors.toList());
+      final JsonNode properties = streamSchema.findValue("properties");
+      if (properties.isObject()) {
+        ((ObjectNode) properties).retain(selectedFieldNames);
```

Comment:
what if there's a selected field that doesn't appear in properties? i think i need to pull this down to run locally to fully understand all this logic and what order stuff happens, but wanted to comment just in case we're missing something

Reply:
Good question. We could tolerate this but I think it's better to throw an error; adding in the latest revision.